### PR TITLE
Add indexOf method

### PR DESCRIPTION
**Pull request by geloescht**
This PR adds the method indexOf that is also available for Buffers in node. I implemented two variants: A search for single byte values only, that should be fairly efficient because it just calls indexOf on each buffer that makes up the BufferList, and a search for multi-byte values that span several buffers within the list. The second variant has just been added to provide compatibility with Buffer.indexOf and is not necessarily efficient enough for complex tasks.
I also added rudimentary tests for both implementations.

I added a helper method _index that does just the inverse of _offset and could prove useful elsewhere too.

### CODE REVIEW

**Comment by geloescht** [comment]
Ooops, missed that Buffer.indexOf is only available since node v4. What is the official stance on compatibility with older versions of node?

**Comment by mcollina** [comment]
Well there is no one, but I think we will support 0.10 and 0.12 until they are supported versions.
 Is there a ponyfill for 0.10 and 0.12?

**Comment by rvagg** [comment]
Yeah, we'll need 0.10 and 0.12 support till EOY unfortunately. I'm not aware of a good polyfill and the encoding makes it kind of tricky to get right.

@mcollina what do you think about doing a feature test and only adding the method if `Buffer.prototype.indexOf` exists? then it would only be available on v4+, we could make a `throw`ing alternative for < v4. I don't think I'd mind such an approach cause it's essentially what you get when you try to use `BufferList.prototype.indexOf` directly.

**Comment by mcollina** [comment]
@rvagg I'm 👍 for a feature test and throwing in node < 4.

**Comment by geloescht** [comment]
Ok, I can do it next week :)

**Comment by geloescht** [comment]
Please review.

File: README.md [hunk 1]
```diff
@@ -163,6 +164,13 @@ console.log(bl.toString())
 `consume()` will shift bytes *off the start of the list*. The number of bytes consumed don't need to line up with the sizes of the internal Buffers&mdash;initial offsets will be calculated accordingly in order to give you a consistent view of the data.
 
 --------------------------------------------------------
+<a name="indexOf"></a>
+### bl.indexOf(value, [ byteOffset [, encoding ]])
+`indexOf()` looks for the first occurence of `value` within the list. It will return the starting index position of `value` or `-1` if the list does not contain `value`. The `value` can be a String, Buffer or Number. `byteOffset` specifies the index within list at which to start the search and defaults to `0`. Strings are by default interpreted as UTF8 or as `encoding` if specified.
+
+Note that `indexOf()` operates in a much more efficient fashion if `value` is only one byte long.
```
Comment (mcollina) [inline README.md]: Can you please specify that this is only available in node v4?

**Comment by mcollina** [comment]
The code is somewhat complex, and I fear there are not enough unit tests to cover it. https://github.com/nodejs/node/blob/master/test/parallel/test-buffer-indexof.js has plenty more assertions, especially on strings, encodings etc... can you please add some more of those?

**Comment by geloescht** [comment]
I am hesitant to add tests on encodings myself, because it is a complex topic that I know very little about.

**Comment by mcollina** [comment]
The problem with encoding is multi-byte support, as I have been bitten very hard by those I'm hesitant in adding this functionality without tests covering them: we usually forget to support chinese and other non-western alphabets. How about you just port a bunch of those https://github.com/nodejs/node/blob/master/test/parallel/test-buffer-indexof.js#L127-L179?

**Comment by mcollina** [comment]
Closing for no activity.

**Pull request closed without merging**